packages/experimental/src/inbox-note/inbox-note.tsx

```diff
@@ -0,0 +1,300 @@
+			}
+
+			hasBeenSeen.current = true;
+		}
+	};
+
+	const handleBlur = ( event: React.FocusEvent, onClose: () => void ) => {
+		const dropdownClasses = [
+			'woocommerce-admin-dismiss-notification',
+			'components-popover__content',
+		];
+		// This line is for IE compatibility.
+		let relatedTarget: EventTarget | Element | null = null;
+		if ( event.relatedTarget ) {
+			relatedTarget = event.relatedTarget;
+		} else if ( toggleButtonRef.current ) {
+			const ownerDoc = toggleButtonRef.current.ownerDocument;
+			relatedTarget = ownerDoc ? ownerDoc.activeElement : null;
+		}
+		let isClickOutsideDropdown = false;
+		if ( relatedTarget && 'className' in relatedTarget ) {
+			const classNames = relatedTarget.className;
+			isClickOutsideDropdown = dropdownClasses.some( ( className ) =>
+				classNames.includes( className )
+			);
+		}
+		if ( isClickOutsideDropdown ) {
+			event.preventDefault();
+		} else {
+			onClose();
+		}
+	};
+
+	const onDropdownDismiss = (
+		type: 'note' | 'all',
+		onToggle: () => void
+	) => {
+		if ( onDismiss ) {
+			onDismiss( note, type );
+		}
+		onToggle();
+	};
+
+	const renderDismissButton = () => {
+		if ( clickedActionText ) {
+			return null;
+		}
+
+		return (
+			<DropdownWithPopoverProps
+				contentClassName="woocommerce-admin-dismiss-dropdown"
+				position="bottom right"
+				renderToggle={ ( { onClose, onToggle } ) => (
+					<Button
+						isTertiary
+						onClick={ onToggle }
+						ref={ toggleButtonRef }
+						onBlur={ ( event: React.FocusEvent ) =>
+							handleBlur( event, onClose )
+						}
+					>
+						{ __( 'Dismiss', 'woocommerce-admin' ) }
+					</Button>
+				) }
+				focusOnMount={ false }
+				popoverProps={ { noArrow: true } }
+				renderContent={ ( { onToggle } ) => (
+					<ul>
+						<li>
+							<Button
+								className="woocommerce-admin-dismiss-notification"
+								onClick={ () =>
+									onDropdownDismiss( 'note', onToggle )
+								}
+							>
+								{ __(
+									'Dismiss this message',
+									'woocommerce-admin'
+								) }
+							</Button>
+						</li>
+						<li>
+							<Button
+								className="woocommerce-admin-dismiss-notification"
+								onClick={ () =>
+									onDropdownDismiss( 'all', onToggle )
+								}
+							>
+								{ __(
+									'Dismiss all messages',
+									'woocommerce-admin'
+								) }
+							</Button>
+						</li>
+					</ul>
+				) }
+			/>
+		);
+	};
+
+	const onActionClicked = ( action: InboxNoteAction ) => {
+		if ( onNoteActionClick ) {
+			onNoteActionClick( note, action );
+		}
+		if ( ! action.actioned_text ) {
+			return;
+		}
+
+		setClickedActionText( action.actioned_text );
+	};
+
+	const renderActions = () => {
+		const { actions: noteActions, id: noteId } = note;
+
+		if ( !! clickedActionText ) {
+			return clickedActionText;
+		}
+
+		if ( ! noteActions ) {
+			return;
+		}
+
+		return (
+			<>
+				{ noteActions.map( ( action, index ) => (
+					<InboxNoteActionButton
+						key={ index }
```

Comment:
We shouldn't use the `index` as a key, it leads to strange bugs, and as a result is just a general anti-pattern. Again I know this is probably just extracted code so up to you if you fix.

Probably a stable ID could be generated from some combo of label and url or something.